### PR TITLE
SIR-887 : Story – WP – Check your answers

### PR DESCRIPTION
**Pull request by sujithvg**

### CODE REVIEW

File: server/routes/__tests__/water-pollution/check-your-answers.spec.js [hunk 1]
```diff
@@ -0,0 +1,490 @@
...
+    })
+    it(`Happy: Should return correct answer for 'Is there anything else you'd like to add?' question ${url}`, async () => {
+      const answerData = {
+        'water-pollution/other-information': 'Details of other information'
+      }
+      sessionData = { ...sessionData, ...answerData }
+      const response = await submitGetRequest({ url }, header, constants.statusCodes.OK, sessionData)
+      expect(response.payload).toContain('Details of other information')
+    })
+    /* it(`Happy: Should return correct answer 'No' if there is no input for 'Is there anything else you'd like to add?' question ${url}`, async () => {
```
Comment (teddmason) [inline server/routes/__tests__/water-pollution/check-your-answers.spec.js:354]: remove commented code

File: server/routes/__tests__/water-pollution/check-your-answers.spec.js [hunk 2]
```diff
@@ -0,0 +1,490 @@
...
+      sessionData = { ...sessionData, ...answerData }
+      const response = await submitGetRequest({ url }, header, constants.statusCodes.OK, sessionData)
+      expect(response.payload).toContain('4th October 2024 at 09:09 AM')
+    })
+    it(`Happy: Should return correct answer with date suffix default case 'th' for 'When did you see the pollution?' question ${url}`, async () => {
+      const answerData = {
+        'water-pollution/when': '2024-10-25T08:09:00.000Z'
+      }
+      sessionData = { ...sessionData, ...answerData }
+      const response = await submitGetRequest({ url }, header, constants.statusCodes.OK, sessionData)
```
Comment (teddmason) [inline server/routes/__tests__/water-pollution/check-your-answers.spec.js:242]: For dates tests, it looks like all of the tests with fixed dates are for BST, would be good to get one for GMT too.

File: server/routes/__tests__/water-pollution/other-information.spec.js [hunk 3]
```diff
@@ -23,125 +20,8 @@ describe(url, () => {
...
-              answerId: 103,
-              otherDetails: 'other details'
-            })
-          ])
-        })
-      }))
+      const response = await submitPostRequest(options)
       expect(response.request.yar.get(constants.redisKeys.WATER_POLLUTION_OTHER_INFORMATION)).toEqual(otherInfo)
-      expect(new Date(response.request.yar.get(constants.redisKeys.SUBMISSION_TIMESTAMP))).toBeInstanceOf(Date)
-      expect(response.headers.location).toEqual(constants.routes.REPORT_SENT)
```
Comment (teddmason) [inline server/routes/__tests__/water-pollution/other-information.spec.js:128]: Please can you add this check back in to check the headers.location matches the check your answers route

**Review by teddmason — CHANGES_REQUESTED** [review]
A few comments to look through, no major issues

File: server/routes/water-pollution/check-your-answers.js [hunk 4]
```diff
@@ -0,0 +1,370 @@
...
+  }
+}
+
+// Get answers for 'Location and size of pollution' section
+const getLocationAndSizeOfPollution = (request) => {
+  // Get answer for 'Type of water' question
+  const waterFeatureUrl = 'WATER_POLLUTION_WATER_FEATURE'
+  const waterFeatureAnswer = getDataSet(request, waterFeatureUrl)
+
+  console.log('Data for waterFeatureAnswer', waterFeatureAnswer)
```
[the remark this answers is not in the document]
Comment (sujithvg) [reply server/routes/water-pollution/check-your-answers.js]: Removed the console logs

File: server/routes/water-pollution/check-your-answers.js [hunk 5]
```diff
@@ -0,0 +1,370 @@
...
+
+  // Get answer for 'Less than 10m in size' question
+  const lessThan10MetersUrl = 'WATER_POLLUTION_LESS_THAN_10_METRES'
+  const lessThan10MetersAnswer = getData(request, lessThan10MetersUrl)
+
+  // Get answer for 'Less than 100 square meters in size' question
+  const lessThan100SqMetersUrl = 'WATER_POLLUTION_LESS_THAN_100_SQ_METRES'
+  const lessThan100SqMetersAnswer = getData(request, lessThan100SqMetersUrl)
+
+  console.log('Data for lessThan100SqMetersAnswer', lessThan100SqMetersAnswer)
```
[the remark this answers is not in the document]
Comment (sujithvg) [reply server/routes/water-pollution/check-your-answers.js]: Removed the console logs

File: server/routes/water-pollution/check-your-answers.js [hunk 6]
```diff
@@ -0,0 +1,370 @@
...
+
+  // Get answer for 'Size (estimated)' question
+  if (isSizeEstimatedRequired) {
+    const pollutionLengthURL = 'WATER_POLLUTION_POLLUTION_LENGTH'
+    const pollutionLengthAnswer = getData(request, pollutionLengthURL)
+
+    const pollutionAreaURL = 'WATER_POLLUTION_POLLUTION_AREA'
+    const pollutionAreaAnswer = getData(request, pollutionAreaURL)
+
+    if (pollutionLengthAnswer !== null && pollutionLengthAnswer) {
```
[the remark this answers is not in the document]
Comment (sujithvg) [reply server/routes/water-pollution/check-your-answers.js]: Changed the code as per the comments 👍

File: server/routes/water-pollution/check-your-answers.js [hunk 7]
```diff
@@ -0,0 +1,370 @@
...
+      sizeEstimatedAnswer = pollutionLengthAnswer
+    } else if (pollutionAreaAnswer !== null && pollutionAreaAnswer) {
```
[the remark this answers is not in the document]
Comment (sujithvg) [reply server/routes/water-pollution/check-your-answers.js]: Changed the code as per the comments 👍

File: server/routes/water-pollution/check-your-answers.js [hunk 8]
```diff
@@ -0,0 +1,370 @@
...
+}
+
+// Get answers for 'Your details' section
+const getYourDetails = (request) => {
+  // Get answer for 'Name, Phone number and Email address' questions
+  const { reporterName, reporterPhoneNumber, reporterEmailAddress } = request.yar.get(constants.redisKeys.HOME)
+
+  // Get answer for 'Images or videos available' question
+  const imagesOrVideoUrl = 'WATER_POLLUTION_IMAGES_OR_VIDEO'
+  const imagesOrVideoAnswer = getData(request, imagesOrVideoUrl)
```
[the remark this answers is not in the document]
Comment (sujithvg) [reply server/routes/water-pollution/check-your-answers.js:47]: Removed all the additional declarations and changed the code as per the comments 👍

File: server/routes/water-pollution/check-your-answers.js [hunk 9]
```diff
@@ -0,0 +1,370 @@
...
+  const effectOnWildlifeAnswer = getDataSet(request, effectOnWildlifeUrl)
+
+  // Get answer for 'Is there anything else you'd like to add?' question
+  const otherInformationUrl = 'WATER_POLLUTION_OTHER_INFORMATION'
+  const otherInformationAnswerData = request.yar.get(constants.redisKeys[otherInformationUrl])
+  let otherInformationAnswer
+
+  if (otherInformationAnswer !== null && otherInformationAnswerData) {
+    otherInformationAnswer = otherInformationAnswerData
+  }
```
[the remark this answers is not in the document]
Comment (sujithvg) [reply server/routes/water-pollution/check-your-answers.js]: Changed the code as per the comments 👍

File: server/routes/water-pollution/check-your-answers.js [hunk 10]
```diff
@@ -0,0 +1,370 @@
...
+    pollutionSourceAnswer,
+    effectOnWildlifeAnswer,
+    otherInformationAnswer
+  }
+}
+
+// Get data and construct answers for the questions
+const getData = (request, pageUrl) => {
+  const recordedAnswer = request.yar.get(constants.redisKeys[pageUrl])
+  console.log('Data for recordedAnswer', recordedAnswer)
```
[the remark this answers is not in the document]
Comment (sujithvg) [reply server/routes/water-pollution/check-your-answers.js]: Removed console logs

File: server/routes/water-pollution/check-your-answers.js [hunk 11]
```diff
@@ -0,0 +1,370 @@
...
+  if (recordedAnswer !== null && recordedAnswer.length === 1) {
```
[the remark this answers is not in the document]
Comment (sujithvg) [reply server/routes/water-pollution/check-your-answers.js]: Changed the code as per the comments 👍

File: server/routes/water-pollution/check-your-answers.js [hunk 12]
```diff
@@ -0,0 +1,370 @@
...
+  } else {
+    return null
+  }
+}
+
+// Get data and construct multiple answers for the questions
+const getDataSet = (request, pageUrl) => {
+  const recordedAnswerSet = getData(request, pageUrl)
+  let answerData
+  if (recordedAnswerSet !== null && recordedAnswerSet.otherDetails) {
```
[the remark this answers is not in the document]
Comment (sujithvg) [reply server/routes/water-pollution/check-your-answers.js]: Changed the code as per the comments 👍

File: server/routes/water-pollution/check-your-answers.js [hunk 13]
```diff
@@ -0,0 +1,370 @@
...
+    const joinData = recordedAnswerSet.answerText.join('<br>')
+    answerData = `${joinData} - ${recordedAnswerSet.otherDetails}`
+  } else if (recordedAnswerSet !== null && recordedAnswerSet.answerText) {
```
[the remark this answers is not in the document]
Comment (sujithvg) [reply server/routes/water-pollution/check-your-answers.js]: Changed the code as per the comments 👍

File: server/routes/__tests__/water-pollution/check-your-answers.spec.js (hunk 1 above)
Comment (sujithvg) [reply server/routes/__tests__/water-pollution/check-your-answers.spec.js]: done

File: server/routes/__tests__/water-pollution/check-your-answers.spec.js (hunk 2 above)
Comment (sujithvg) [reply server/routes/__tests__/water-pollution/check-your-answers.spec.js]: Added GMT tests

File: server/routes/__tests__/water-pollution/other-information.spec.js (hunk 3 above)
Comment (sujithvg) [reply server/routes/__tests__/water-pollution/other-information.spec.js:128]: added back in the headers.location and pointed it to the CYA page

File: server/routes/water-pollution/check-your-answers.js [hunk 14]
```diff
@@ -0,0 +1,370 @@
...
+  return isSizeEstimatedRequiredData
+}
+
+// Format date and time data
+const getWhenData = (request, pageUrl) => {
+  const whenAnswerData = request.yar.get(constants.redisKeys[pageUrl])
+  let dateTimeData
+  const pollutionDateAndTime = new Date(whenAnswerData)
+
+  if (whenAnswerData !== null && whenAnswerData) {
```
[the remark this answers is not in the document]
Comment (sujithvg) [reply server/routes/water-pollution/check-your-answers.js]: Changed the code as per the comments 👍

**Review by teddmason — APPROVED** [review]

**Pull request merged**